compiler/rustc_index/src/bit_set.rs

```diff
@@ -208,33 +226,167 @@ impl<T: Idx> BitSet<T> {
+            HybridBitSet::Dense(dense) => self.union(dense),
+        }
+    }
+
+    fn subtract(&mut self, other: &HybridBitSet<T>) -> bool {
+        match other {
+            HybridBitSet::Sparse(sparse) => self.subtract(sparse),
+            HybridBitSet::Dense(dense) => self.subtract(dense),
+        }
+    }
+
+    fn intersect(&mut self, other: &HybridBitSet<T>) -> bool {
+        match other {
+            HybridBitSet::Sparse(sparse) => self.intersect(sparse),
+            HybridBitSet::Dense(dense) => self.intersect(dense),
+        }
+    }
+}
+
+impl<T: Idx> BitRelations<HybridBitSet<T>> for HybridBitSet<T> {
+    fn union(&mut self, other: &HybridBitSet<T>) -> bool {
+        match self {
+            HybridBitSet::Sparse(self_sparse) => {
+                match other {
+                    HybridBitSet::Sparse(other_sparse) => self_sparse.union(other_sparse),
+
+                    HybridBitSet::Dense(other_dense) => {
+                        // `self` is sparse and `other` is dense. To
+                        // merge them, we have two available strategies:
+                        // * Densify `self` then merge other
+                        // * Clone other then integrate bits from `self`
+                        // The second strategy requires dedicated method
+                        // since the usual `union` returns the wrong
+                        // result. In the dedicated case the computation
+                        // is slightly faster if the bits of the sparse
+                        // bitset map to only few words of the dense
+                        // representation, i.e. indices are near each
+                        // other.
+                        //
+                        // Benchmarking seems to suggest that the second
+                        // option is worth it.
+                        let mut new_dense = other_dense.clone();
+                        let changed = new_dense.reverse_union_sparse(self_sparse);
+                        *self = HybridBitSet::Dense(new_dense);
+                        changed
+                    }
+                }
+            }
+
+            HybridBitSet::Dense(self_dense) => self_dense.union(other),
+        }
+    }
+
+    fn subtract(&mut self, other: &HybridBitSet<T>) -> bool {
+        // FIXME(willcrichton): should there be an optimized sparse / dense version?
+        match self {
+            HybridBitSet::Sparse(self_sparse) => self_sparse.subtract(other),
+            HybridBitSet::Dense(self_dense) => self_dense.subtract(other),
+        }
+    }
+
+    fn intersect(&mut self, other: &HybridBitSet<T>) -> bool {
+        // FIXME(willcrichton): should there be an optimized sparse / dense version?
+        match self {
+            HybridBitSet::Sparse(self_sparse) => self_sparse.intersect(other),
+            HybridBitSet::Dense(self_dense) => {
+                <BitSet<T> as BitRelations<HybridBitSet<T>>>::intersect(self_dense, other)
```

Comment:
Is the UFCS significant here? Might be good to leave a comment if so. Collisions are why the old helper traits had different names than the inherent methods.

Regarding the FIXMEs, the only obvious potential optimization is when we're intersecting a dense bitset with a sparse one. Since we know the end result will be no bigger than the sparse bitset, we could use it to replace `self` before intersecting it with the dense bitset. You'll be fighting the borrow checker with this one, so no need to do it now if you don't want.